lib/winappdriver.js

```diff
@@ -1,36 +1,250 @@
+    if (!await fs.exists(binPath)) {
+      throw new Error(`Trying to use a WinAppDriver binary at the path ` +
+                      `${binPath}, but it doesn't exist!`);
+    }
+    this.winappdriverpath = binPath;
+    this.executableVerified = true;
+    log.info(`Set WinAppDriver binary as: ${this.winappdriverpath}`);
+  }
+
+  async start (caps, emitStartingState = true) {
+
+    this.capabilities = caps;
+    if (emitStartingState) {
+      this.changeState(WinAppDriver.STATE_STARTING);
+    }
+
+    // XXXYD TODO: would be better if WinAppDriver didn't require passing in /wd/hub as a param
+    let args = [this.proxyPort + "/wd/hub"];
+
+    if (this.cmdArgs) {
+      args = args.concat(this.cmdArgs);
+    }
+
+    const startDetector = (stdout) => {
+        // XXXYD What is this for, looking for a log entry that shows we're up?
+        // return stdout.indexOf('ENTER') != -1;
+        return true; 
+      };
+
+    let processIsAlive = false;
+        try {
+          await this.initWinAppDriverPath();
+          await this.killAll();
+
+          // set up our subprocess object
+          this.proc = new SubProcess(this.winappdriverpath, args);
+          processIsAlive = true;
+
+          // handle log output
+          this.proc.on('output', (stdout, stderr) => {
+            if (stdout) {
+              log.info(`[STDOUT] ${stdout.trim()}`);
+            }
+            if (stderr) {
+              log.info(`[STDERR] ${stderr.trim()}`);
+            }
+          });
+
+     // handle out-of-bound exit by simply emitting a stopped state
+      this.proc.on('exit', (code, signal) => {
+        processIsAlive = false;
+        if (this.state !== WinAppDriver.STATE_STOPPED &&
+            this.state !== WinAppDriver.STATE_STOPPING &&
+            this.state !== WinAppDriver.STATE_RESTARTING) {
+          let msg = `WinAppDriver exited unexpectedly with code ${code}, ` +
+                    `signal ${signal}`;
+          log.error(msg);
+          this.changeState(WinAppDriver.STATE_STOPPED);
+        }
+      });
+      log.info(`Spawning winappdriver with: ${this.winappdriver} ` +
+               `${args.join(' ')}`);
+      // start subproc and wait for startDetector
+      await this.proc.start(startDetector);
+
+      // XXXYD TODO: we need to handle status in WinAppDriver correctly in order to bring this wait back
+      // await this.waitForOnline(); 
+      await this.startSession();
+    } catch (e) {
+      this.emit(WinAppDriver.EVENT_ERROR, e);
+      // just because we had an error doesn't mean the chromedriver process
+      // finished; we should clean up if necessary
+      if (processIsAlive) {
+        await this.proc.stop();
+      }
+      log.errorAndThrow(e);
+    }
+  }      
+
+  sessionId () {
+    if (this.state !== WinAppDriver.STATE_ONLINE) {
+      return null;
+    }
+
+    return this.jwproxy.sessionId;
+  }
+
+  async restart () {
```

Comment:
unclear whether we'll ever need this for winappdriver. we did because of specific chromedriver issues.